Added a way to check if players are abusing it to cheat death

Ghoulboy78 opened the pull request:
Basically, it won't work if you are in or above lava or fire, and if you are falling through the air. Please tell me if I should add other things.

Ghoulboy78-Alt committed:
Added scoreboard 'valid'

It is used when changing to spectator to see if it's cheating death or not.

Ghoulboy78-Alt committed:
Added conditions for changing to spectator

This will make sure players aren't abusing it to cheat death.
Current options: in lava/fire, falling, falling into fire/lava
BUT it will allow you if you have no FallDistance, i.e you are crouching on a a block or something and are supported by it even though technically you are in another block, but you are not falling.

Ghoulboy78-Alt committed:
Added conditions for changing to spectator

This will make sure players aren't abusing it to cheat death.
Current options: in lava/fire, falling, falling into fire/lava
BUT it will allow you if you have no FallDistance, i.e you are crouching on a a block or something and are supported by it even though technically you are in another block, but you are not falling.

Ghoulboy78-Alt committed:
Added cactus

If you are above cactus, you can't go into spectator. But I didn't add to the side, as you could be in a desert, and cactus is less than a block wide, and I don't know how to check that specifically. Feel free to add it if you know how

Ghoulboy78-Alt committed:
Imporved the lava/fire bit

Now it will also prevent you from going into spec mode if your head block or the block above is on fire.

Ghoulboy78-Alt committed:
Added description of changes to README.MD

Ghoulboy78-Alt committed:
Added drowning damage detection

Pablete1234 commented on spectator/data/spectator/functions/spectator.mcfunction:
```diff
@@ -1,8 +1,23 @@
+execute if block ~ ~1 ~ minecraft:lava run scoreboard players set @s valid 0
+execute if block ~ ~1 ~ minecraft:fire run scoreboard players set @s valid 0
+execute if block ~ ~2 ~ minecraft:lava run scoreboard players set @s valid 0
+execute if block ~ ~2 ~ minecraft:fire run scoreboard players set @s valid 0
```
Could just use the Fire NBT tag instead. It should be at -20 if you're not in lava nor fire. I don't think it's worth checking so many blocks

Pablete1234 commented on spectator/data/spectator/functions/spectator.mcfunction:
```diff
@@ -1,8 +1,23 @@
+execute if block ~ ~2 ~ minecraft:lava run scoreboard players set @s valid 0
+execute if block ~ ~2 ~ minecraft:fire run scoreboard players set @s valid 0
+execute if block ~ ~-1 ~ minecraft:air unless data entity @s {FallDistance:0f} run scoreboard players set @s valid 0
+execute if block ~ ~-1 ~ minecraft:cactus run scoreboard players set @s valid 0
```
I think this one isn't really needed. Being ontop of a cactus will rarely kill you and it's easier to step off one than to trigger into spectator

Pablete1234 commented on spectator/data/spectator/functions/spectator.mcfunction:
```diff
@@ -1,8 +1,23 @@
+execute if block ~ ~2 ~ minecraft:fire run scoreboard players set @s valid 0
+execute if block ~ ~-1 ~ minecraft:air unless data entity @s {FallDistance:0f} run scoreboard players set @s valid 0
+execute if block ~ ~-1 ~ minecraft:cactus run scoreboard players set @s valid 0
+execute if data entity @s {Air:0s} run scoreboard players set @s valid 0
```
I think it's better to enforce air=300. If you're under water you shouldn't be triggering spectator. Recover your air  first.

Pablete1234 commented on spectator/data/spectator/functions/spectator.mcfunction:
```diff
@@ -1,8 +1,23 @@
+execute if block ~ ~1 ~ minecraft:fire run scoreboard players set @s valid 0
+execute if block ~ ~2 ~ minecraft:lava run scoreboard players set @s valid 0
+execute if block ~ ~2 ~ minecraft:fire run scoreboard players set @s valid 0
+execute if block ~ ~-1 ~ minecraft:air unless data entity @s {FallDistance:0f} run scoreboard players set @s valid 0
```
If you're trapped in a cobweb (potentially going to die from fall or lava later), FallDistance is 0. Consider using OnGround instead.
You'll need to check for OnGround OR current block (or block under you) is water.

Pablete1234 reviewed (changes requested):
Checking for danger or inminent death is generally a very hard thing, there are so many things that could put you in danger that it's impossible to check for all of them, but we can definetly check for falling, fire and drawning.

You'll also need to inform the player that they're in too much danger to get into spectator, otherwise they'll think the datapack is broken if they try to use it under water or something similar

Ghoulboy78-Alt committed:
Improved fire damage to be more efficient, and all the other stuff Pablo said

Ghoulboy78-Alt committed:
Added feedback command

That way players will know that they are in an invalid location, and not that the datapack is broken

Ghoulboy78 commented:
It think I've done all those changes @Pablete1234 , please review and tell me if I've done something wrong.
Also, should I make it so you can't do it if you are near mobs? Cos otherwise you could use it to escape mob fights.

Pablete1234 commented:
Looking for entity data is going to get annoying & give potential falsle positives, they could be near trapped mobs that are irrelevant

Pablete1234 commented on spectator/data/spectator/functions/spectator.mcfunction:
```diff
@@ -1,8 +1,16 @@
\ No newline at end of file
+#Conditions for cheating (block is fire/lava, or falling)
+execute unless data entity @s {Air:300s} run scoreboard players set @s valid 0
+execute unless data entity @s {OnGround:1b} run scoreboard players set @s valid 0
```
This won't let you go into spectator if floating on water. If you passed the Air:300 check this should allow for water

Ghoulboy78 replied:
and exactly how do you float on water?

Pablete1234 replied:
Pressing spacebar

Ghoulboy78 replied:
but how do you even enter the command while having spacebar held down? cos when you press T to open chat, you begin to sink

Pablete1234 replied:
You sink slower than you can type, and the command is ran up to 1s later than you entering it. You have the time it takes to sink from feet to head to get into spec, any later will get you denied for Air time (drowning), but the point is being on water should pass the "on ground" check

Ghoulboy78 replied:
@Pablete1234 Please view commit 0ae9ad4 on 12th march to see if that's what you mean

Ghoulboy78 commented:
Also, about the mob thing, that's why I asked, cos I thought about whether it could lead to unwanted results. Also,should we check if you're flying on an elytra?

Ghoulboy78-Alt committed:
I hope this is what you meant, Pablo, by being in water and not drowning

This way, if you're in water, and not drowning, it will first invalidate it for not being on ground, then revalidate it for being in water, but if you're like in 1-block deep, your air will be 300, so you can still do it. But if you're drowning, it will invalidate it for not being on land, then revalidate it for being in water, then reinvalidate it for drowning

Ghoulboy78-Alt committed:
Made it more efficient
